### PR TITLE
Introduce loadable and savable HOC

### PR DESCRIPTION
**Pull request by viroulep**
As we progressively redesign the website or simply implement new frontend parts in React, we will need a way of asynchronously loading and saving objects from the WCA website's json endpoints.
I tried to think of something generic enough, and I came up with these two HOC that can wrap components and provide them with loaded data and save functions.
The id provided can be either a string or an object: it's oblivious to the HOC since it passes it to the route helper, which knows how to handle the id.
You can also note that without the HOC, wrapped components behave just fine (as long as they use the same id + loadedState props to fill their state).

I know I didn't include any component using it yet, but that's an attempt to make the review of some upcoming works more easy.
If you want to see it in action you can take a look at what I started to do for an "add/edit results" form for the WRT [here](https://github.com/viroulep/worldcubeassociation.org/blob/d11fafdfa8bcec831139251d5a16d255a80436d7/WcaOnRails/app/javascript/edit-results/EditResult/ResultForm.jsx#L17) (there are a bit more usage, you need to dig in the 'EditResult' directory.

Recently I also used it to implement online competitions on the French website, you can see it live [here](https://www.speedcubingfrance.org/online_competitions/ConfinementOpen3) for instance (it's used to load the results, or load/save user submitted results).
[Here](https://github.com/speedcubingfrance/speedcubingfrance.org/blob/27f2a7f9c27dd2c4c7b70a7cfbefe25002133560/app/webpacker/results/LoadableResult.js#L9-L31) is how the code looks like for a loadable *and* savable component.

You can see how I designed the routes to work [here](https://github.com/speedcubingfrance/speedcubingfrance.org/blob/27f2a7f9c27dd2c4c7b70a7cfbefe25002133560/app/webpacker/requests/routes.js.erb#L5), with an example of non-basic id object.

@jonatanklosko I'm very interested in having your feedback about this, since I basically figured this on my own, trying to address the "How can I make our components loadable and savable, without being intrusive, and while being able to query our rails controller in a json format?" question.

### CODE REVIEW

File: WcaOnRails/app/javascript/requests/savable.jsx [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+//   - urlHelper is a function which takes one parameter: an identifier of the
+//   resource to set.
+//   - 'id' is passed in the props, and will be fed to the urlHelper.
+//   For new record, 'id' may be null, urlHelper is responsible for returning
+//   the appropriate url.
+//   - 'saveState' is pass down to the wrapped component, and can be used to
+//   save the value (one param: data).
+//   - 'saveErrors' is pass down to the wrapped component if any error arised.
+// Example of usage:
+// const ModelComponent = savableComponent(id => { /* component stuff */ }, id => `path/to/resource/${id}` || `path/to/new/resource`;);
```
Comment (jonatanklosko) [inline WcaOnRails/app/javascript/requests/savable.jsx]: Same note to `id => { /* component stuff */ }` as before.

I believe the `urlHelper` should be `id => id ? ... : ...`.

File: WcaOnRails/app/javascript/requests/savable.jsx [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+      };
+    }
+
+    saveState = (data, onSuccess) => {
+      this.setState({ saving: true });
+      fetchJsonOrError(urlHelper(this.props.id), {
+        headers: {
+          "Content-Type": "application/json",
+        },
+        //credentials: 'include',
```
Comment (jonatanklosko) [inline WcaOnRails/app/javascript/requests/savable.jsx]: I guess that was temporary?

File: WcaOnRails/app/javascript/requests/savable.jsx [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+// This is a HOC that can be used to save a data to the website (as json)
+// It assumes that:
+//   - urlHelper is a function which takes one parameter: an identifier of the
+//   resource to set.
+//   - 'id' is passed in the props, and will be fed to the urlHelper.
+//   For new record, 'id' may be null, urlHelper is responsible for returning
+//   the appropriate url.
+//   - 'saveState' is pass down to the wrapped component, and can be used to
+//   save the value (one param: data).
+//   - 'saveErrors' is pass down to the wrapped component if any error arised.
```
Comment (jonatanklosko) [inline WcaOnRails/app/javascript/requests/savable.jsx]: I cannot see that in the implementation, am I missing something?

File: WcaOnRails/app/javascript/requests/loadable.jsx [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    }
+
+    componentDidUpdate(prevProps) {
+      if (this.props.id != prevProps.id) {
+        this.loadState();
+      }
+    }
+
+    render() {
+      return <WrappedComponent {...this.props} loadedState={this.state.loadedState || this.props.loadedState} />;
```
[the remark this answers is not in the document]
Comment (viroulep) [reply WcaOnRails/app/javascript/requests/loadable.jsx]: Ha that's a great idea!

File: WcaOnRails/app/javascript/requests/savable.jsx (hunk 3 above)
Comment (viroulep) [reply WcaOnRails/app/javascript/requests/savable.jsx]: You're missing my first attempts which were a bit clunky and did not convince me, I totally forgot to clean up the comment :)

**Comment by viroulep** [comment]
Thanks a lot for your comments @jonatanklosko!
I'll try your suggestions and update the PR with them, as well as cleanup the inaccurate comments.

**Comment by viroulep** [comment]
I actually found a much better and less intrusive version of doing that, by only using react hooks.
I've added this to #5340 but I'll leave the code here too in case someone comes upon this PR later:
```javascript
import { useState, useEffect, useCallback } from 'react';
import { fetchJsonOrError } from './fetchWithAuthenticityToken';

// This is a hook that can be used to get a data from the website (as json)
// It assumes that 'url' is a valid, GET-able, url.
// Example of usage:
// const { loadedData, loading, error, sync } = useLoadedData(`path/to/resource`);
const useLoadedData = (url) => {
  const [loadedData, setData] = useState(null);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(null);

  const sync = useCallback(() => {
    setLoading(true);
    setError(null);
    fetchJsonOrError(url).then((loaded) => {
      setData(loaded);
    }).catch((err) => {
      setError(err.message);
    }).finally(() => setLoading(false));
  }, [url, setData, setError]);

  useEffect(sync, [sync]);

  return {
    loadedData,
    loading,
    error,
    sync,
  };
};

export default useLoadedData;
```

It provides the user with an accurate view of the loading state, they don't have to remember what to provide and what is provided, and it also provide a way to sync the same resource if needed!

**Comment by jonatanklosko** [comment]
Ah, our current codebase doesn't use hooks much, so I didn't even consider them, that's absolutely perfect! :shipit: :tada:

**Comment by viroulep** [comment]
Closing in favor of #6359.

**Pull request closed without merging**